base/set.jl

```diff
@@ -474,3 +474,220 @@ end
+`pred(x)` is true are replaced by `new`.
+
+# Examples
+```jldoctest
+julia> replace(isodd, [1, 2, 3, 1], 0, count=2)
+4-element Array{Int64,1}:
+ 0
+ 2
+ 0
+ 1
+```
+"""
+replace(pred::Callable, A, new; count::Integer=typemax(Int)) =
+    _replace!(x -> if pred(x) Some(new) end, copy(A), count)
+
+"""
+    replace(prednew::Function, A; [count::Integer])
+
+Return a copy of `A` where for each value `x` in `A`, `prednew(x)` is called
+and must return  either `nothing`, in which case no replacement occurs,
+or a value, possibly wrapped as a [`Some`](@ref) object, which
+will be used as a replacement for `x`.
+
+# Examples
+```jldoctest
+julia> replace(Dict(1=>2, 3=>4)) do kv
+           if first(kv) < 3; first(kv)=>3 end
+       end
+Dict{Int64,Int64} with 2 entries:
+  3 => 4
+  1 => 3
+```
+"""
+replace(prednew::Callable, A; count::Integer=typemax(Int)) = _replace!(prednew, copy(A), count)
+
+# Handle ambiguities
+replace!(a::Callable, b::Pair; count::Integer=-1) = throw(MethodError(replace!, a, b))
+replace!(a::Callable, b::Pair, c::Pair; count::Integer=-1) = throw(MethodError(replace!, a, b, c))
+replace(a::Callable, b::Pair; count::Integer=-1) = throw(MethodError(replace, a, b))
+replace(a::Callable, b::Pair, c::Pair; count::Integer=-1) = throw(MethodError(replace, a, b, c))
+replace(a::AbstractString, b::Pair, c::Pair) = throw(MethodError(replace, a, b, c))
+
+
+### _replace! for AbstractDict/AbstractSet
+
+askey(k, ::AbstractDict) = k.first
+askey(k, ::AbstractSet) = k
+
+function _replace_update_hash!(repl::Vector{<:Pair}, x, y::Some)
+    push!(repl, x => y.value)
+    true
+end
+
+_replace_update_hash!(repl::Vector{<:Pair}, x, ::Void) = false
+_replace_update_hash!(repl::Vector{<:Pair}, x, y) = _replace_update_hash!(repl, x, Some(y))
+
+function _replace!(prednew::Callable, A::Union{AbstractDict,AbstractSet}, count::Int)
+    count < 0 && throw(DomainError(count, "`count` must not be negative"))
+    count == 0 && return A
+    repl = Pair{eltype(A),eltype(A)}[]
+    c = 0
+    for x in A
+        c +=_replace_update_hash!(repl, x, prednew(x))
+        c == count && break
+    end
+    for oldnew in repl
+        pop!(A, askey(first(oldnew), A))
+        push!(A, last(oldnew))
+    end
+    A
+end
+
+### AbstractArray
+
+function _replace_update!(A::AbstractArray, i::Integer, y::Some)
+    @inbounds A[i] = y.value
+    true
+end
+
+_replace_update!(A::AbstractArray, i::Integer, ::Void) = false
+_replace_update!(A::AbstractArray, i::Integer, y) = _replace_update!(A, i, Some(y))
+
+# we use this _replace wrapper because otherwise the performance is
+# degraded when forwarding the keyword argument
+function _replace!(prednew::Callable, A::AbstractArray, count::Int)
+    count < 0 && throw(DomainError(count, "`count` must not be negative"))
+    count == 0 && return A
+    c = 0
+    for i in eachindex(A)
+        c += _replace_update!(A, i, prednew(A[i]))
+        c == count && break
```

Comment:
That's probably an overkill optimization, but I wonder whether using a special loop for `count == typemax(Int)` could be worth it in some cases where the function is so simple that SIMD might be enabled without the `break` call. Maybe leave that to future improvements after some benchmarking.

Reply:
I will add that on my todo list to not forget :)